src/Http/Shared/StreamCopyOperationInternal.cs

```diff
@@ -83,5 +84,66 @@ public static async Task CopyToAsync(Stream source, Stream destination, long? co
                 ArrayPool<byte>.Shared.Return(buffer);
             }
         }
+
+        /// <summary>Asynchronously reads the given number of bytes from the source stream and writes them using pipe writer.</summary>
+        /// <returns>A task that represents the asynchronous copy operation.</returns>
+        /// <param name="source">The stream from which the contents will be copied.</param>
+        /// <param name="writer">The PipeWriter to which the contents of the current stream will be copied.</param>
+        /// <param name="count">The count of bytes to be copied.</param>
+        /// <param name="cancel">The token to monitor for cancellation requests.</param>
+        public static Task CopyToAsync(Stream source, PipeWriter writer, long? count, CancellationToken cancel)
+        {
+            if (count == null)
+            {
+                // No length, do a copy with the default buffer size (based on whatever the pipe settings are, default is 4K)
+                return source.CopyToAsync(writer, cancel);
+            }
+
+            static async Task CopyToAsync(Stream source, PipeWriter writer, long bytesRemaining, CancellationToken cancel)
+            {
+                // The array pool likes powers of 2
+                const int maxBufferSize = 64 * 1024;
+                const int minBufferSize = 1024;
+
+                // We know exactly how much we're going to copy
+                while (bytesRemaining > 0)
+                {
+                    var bufferSize = (int)Math.Clamp(bytesRemaining, minBufferSize, maxBufferSize);
+
+                    // The natural end of the range.
+                    var memory = writer.GetMemory(bufferSize);
+
+                    if (memory.Length > bytesRemaining)
+                    {
+                        memory = memory.Slice(0, (int)bytesRemaining);
+                    }
+
+                    var read = await source.ReadAsync(memory, cancel);
+
+                    bytesRemaining -= read;
+
+                    // End of the source stream.
+                    if (read == 0)
+                    {
+                        break;
```

Comment:
Should you advance here to get the writer in a working state?

Reply:
I think that only matters if you write something. At least in the Pipe implementation.

Reply:
I thought if you call GetMemory, you can't call GetMemory afterwards without advancing.

Reply:
Nope.